src/commands/pull.ts

```diff
@@ -0,0 +1,156 @@
+import type { Blob } from 'buffer';
+import type { Entry } from 'yauzl';
+
+import { join, resolve } from 'path';
+import { createWriteStream } from 'fs';
+import { stat, rm, mkdir } from 'fs/promises';
+import { fromBuffer as zipFromBuffer } from 'yauzl';
+import { Command, Option } from 'commander';
+
+import { API_URL_OPT, API_KEY_OPT, PROJECT_ID_OPT } from '../options';
+import ExportClient from '../client/export';
+
+import { askBoolean } from '../utils/ask';
+import { loading, success, warn, error } from '../utils/logger';
+
+type PullParams = {
+  apiUrl: URL;
+  apiKey: string;
+  projectId: number;
+  format: 'JSON' | 'XLIFF';
+  languages?: string[];
+  states?: Array<'UNTRANSLATED' | 'TRANSLATED' | 'REVIEWED'>;
+  overwrite?: boolean;
+};
+
+async function validatePath(path: string, overwrite?: boolean) {
+  try {
+    const stats = await stat(path);
+    if (!stats.isDirectory()) {
+      error('The specified path already exists and is not a directory.');
+      process.exit(1);
+    }
+
+    if (!overwrite) {
+      if (!process.stdout.isTTY) {
+        error('The specified path already exists.');
+        process.exit(1);
+      }
+
+      warn(`The specified path ${resolve(path)} already exists.`);
+      const userOverwrite = await askBoolean(
+        'Do you want to overwrite data? *BE CAREFUL, ALL THE CONTENTS OF THE DESTINATION FOLDER WILL BE DESTROYED*.'
+      );
+      if (!userOverwrite) {
+        error('Aborting.');
+        process.exit(1);
+      }
+
+      // Purge data as requested.
+      await rm(path, { recursive: true });
+    }
+  } catch (e: any) {
+    if (e.code !== 'ENOENT') {
+      throw e;
+    }
+  }
+}
+
+async function fetchZipBlob(params: PullParams): Promise<Blob> {
+  const client = new ExportClient({
+    apiUrl: params.apiUrl,
+    apiKey: params.apiKey,
+    projectId: params.projectId,
+  });
+
+  return client.export({
+    format: params.format,
+    languages: params.languages,
+    filterState: params.states,
+    zip: true,
+
+    splitByScope: false, // these as marked as required in the API :shrug:
+    splitByScopeDelimiter: '', // these as marked as required in the API :shrug:
+    splitByScopeDepth: 0, // these as marked as required in the API :shrug:
+  });
+}
+
+async function extractZip(zipBlob: Blob, path: string) {
+  const zipAb = await zipBlob.arrayBuffer();
+  return new Promise<void>((resolve, reject) => {
+    const opts = {
+      strictFileNames: true,
+      decodeStrings: true,
+      lazyEntries: true,
+    };
+    zipFromBuffer(Buffer.from(zipAb), opts, (err, zip) => {
+      if (err) return reject(err);
+      zip.on('error', reject);
+
+      zip.readEntry();
+      zip.on('entry', (entry: Entry) => {
+        // Shouldn't be necessary but harmless :shrug:
+        // Directory traversal is protected by yauzl's `decodeStrings`
+        if (entry.fileName.endsWith('/')) {
+          return mkdir(join(path, entry.fileName)).then(() => zip.readEntry());
```

Comment:
await would be nice here, also I was looking into this 2 lines for like 10 minutes to understand what's going on. Can you maybe create separate fn named "handleZipDirectory" or something like that. It would be also cool, since you would remove the comments from function body to a function documentation. Comments in method bodies are pretty vulnerable. Developers can forget to delete them or update them and it creates mess...

Reply:
Using `await` here would introduce a mix of `async/await` and callbacks, which I usually find quite not super clean; I used `.then` to stay in the callback way of doing things.

I'm overall unsure about moving this bit to a separate function. Most of the body of the entry handler is from yauzl's documentation, and having a function just for this, passing the `zip` and `entry` references. The `zip.readEntry()` is equivalent to an iterator's `next` function and it wouldn't make sense to me to call it in a `handleZipDirectory` function, which means only the `mkdir` call would be moved.

I get the remarks about the comments; I haven't really ever thought of comments this way. That being said, I feel like the first comment about the fact this is kind of unnecessary strictly related to the `if` check; the comment implies that this code branch might not even be reached in the first place but deliberately left here. The second comment refers to the use of `fileName` without sanitizing for path traversal on-site, and acknowledges the security problems that may arise from this specific statement and this statement only.

The code can be made cleaner and more explicit, but I have mixed feelings about splitting it off completely 🤔

Reply:
No problem. It was just a tip. To make it super clean, I would create separate class for the ZIP extracting stuff and rewrite it to work with promises. Thanks for clarification.